File: code/packages/python/gpu-core/tests/test_memory.py
"""Tests for LocalMemory."""

from __future__ import annotations

import pytest
from fp_arithmetic import BF16, FP16, FP32, bits_to_float, float_to_bits

from gpu_core.memory import LocalMemory


class TestConstruction:
    """Test memory creation."""

    def test_default_size(self) -> None:
        """Default memory is 4096 bytes."""
        mem = LocalMemory()
        assert mem.size == 4096

    def test_custom_size(self) -> None:
        """Can create memory with custom size."""
        mem = LocalMemory(size=256)
        assert mem.size == 256

    def test_invalid_size(self) -> None:
        """Cannot create memory with size < 1."""
        with pytest.raises(ValueError, match="positive"):
            LocalMemory(size=0)

    def test_initialized_to_zero(self) -> None:
        """Memory starts as all zeros."""
        mem = LocalMemory(size=16)
        for i in range(16):
            assert mem.read_byte(i) == 0


class TestByteAccess:
    """Test raw byte read/write operations."""

    def test_read_write_byte(self) -> None:
        """Write a byte and read it back."""
        mem = LocalMemory()
        mem.write_byte(0, 0x42)
        assert mem.read_byte(0) == 0x42

    def test_byte_masking(self) -> None:
        """Values are masked to 8 bits."""
        mem = LocalMemory()
        mem.write_byte(0, 0x1FF)  # 9 bits
        assert mem.read_byte(0) == 0xFF  # truncated to 8

    def test_read_write_bytes(self) -> None:
        """Write and read multiple bytes."""
        mem = LocalMemory()
        data = b"\x01\x02\x03\x04"
        mem.write_bytes(0, data)
        assert mem.read_bytes(0, 4) == data

    def test_out_of_bounds_read(self) -> None:
        """Reading past memory bounds raises IndexError."""
        mem = LocalMemory(size=8)
        with pytest.raises(IndexError, match="out of bounds"):
            mem.read_byte(8)

    def test_out_of_bounds_write(self) -> None:
        """Writing past memory bounds raises IndexError."""
        mem = LocalMemory(size=8)
        with pytest.raises(IndexError, match="out of bounds"):
            mem.write_byte(8, 0)

    def test_negative_address(self) -> None:
        """Negative addresses are out of bounds."""
        mem = LocalMemory()
        with pytest.raises(IndexError):
            mem.read_byte(-1)

    def test_multi_byte_out_of_bounds(self) -> None:
        """Multi-byte access that extends past the end fails."""
        mem = LocalMemory(size=8)
        with pytest.raises(IndexError):
            mem.read_bytes(6, 4)  # bytes 6,7,8,9 — 8 and 9 out of bounds


class TestFloatAccess:
    """Test floating-point load/store operations."""

    def test_store_load_fp32(self) -> None:
        """Store and load an FP32 value."""
        mem = LocalMemory()
        value = float_to_bits(3.14, FP32)
        mem.store_float(0, value)
        result = mem.load_float(0, FP32)
        assert bits_to_float(result) == pytest.approx(3.14, rel=1e-5)

    def test_store_load_fp16(self) -> None:
        """Store and load an FP16 value (2 bytes)."""
        mem = LocalMemory()
        value = float_to_bits(1.0, FP16)
        mem.store_float(0, value)
        result = mem.load_float(0, FP16)
        assert bits_to_float(result) == 1.0

    def test_store_load_bf16(self) -> None:
        """Store and load a BF16 value (2 bytes)."""
        mem = LocalMemory()
        value = float_to_bits(2.0, BF16)
        mem.store_float(0, value)
        result = mem.load_float(0, BF16)
        assert bits_to_float(result) == 2.0

    def test_fp32_uses_4_bytes(self) -> None:
        """FP32 store writes exactly 4 bytes."""
        mem = LocalMemory()
        value = float_to_bits(1.0, FP32)
        mem.store_float(0, value)
        # Check that bytes 0-3 are non-trivial (1.0 = 0x3F800000)
        raw = mem.read_bytes(0, 4)
        assert len(raw) == 4
        assert raw != b"\x00\x00\x00\x00"

    def test_fp16_uses_2_bytes(self) -> None:
        """FP16 store writes exactly 2 bytes."""
        mem = LocalMemory()
        value = float_to_bits(1.0, FP16)
        mem.store_float(0, value)
        raw = mem.read_bytes(0, 2)
        assert len(raw) == 2
        assert raw != b"\x00\x00"

    def test_multiple_floats_at_different_addresses(self) -> None:
        """Store multiple floats at non-overlapping addresses."""
        mem = LocalMemory()
        mem.store_python_float(0, 1.0)
        mem.store_python_float(4, 2.0)
        mem.store_python_float(8, 3.0)
        assert mem.load_float_as_python(0) == 1.0
        assert mem.load_float_as_python(4) == 2.0
        assert mem.load_float_as_python(8) == 3.0

    def test_store_negative(self) -> None:
        """Store and load a negative float."""
        mem = LocalMemory()
        mem.store_python_float(0, -42.5)
        assert mem.load_float_as_python(0) == -42.5

    def test_store_zero(self) -> None:
        """Store and load zero."""
        mem = LocalMemory()
        mem.store_python_float(0, 0.0)
        assert mem.load_float_as_python(0) == 0.0

    def test_convenience_methods(self) -> None:
        """Test store_python_float and load_float_as_python."""
        mem = LocalMemory()
        mem.store_python_float(0, 2.71828, FP32)
        result = mem.load_float_as_python(0, FP32)
        assert abs(result - 2.71828) < 1e-5

    def test_float_out_of_bounds(self) -> None:
        """Loading a float past memory end raises IndexError."""
        mem = LocalMemory(size=8)
        with pytest.raises(IndexError):
            mem.load_float(6, FP32)  # needs 4 bytes at 6, goes to 10


class TestDump:
    """Test memory dump functionality."""

    def test_dump_zeros(self) -> None:
        """Dump of fresh memory is all zeros."""
        mem = LocalMemory(size=16)
        assert mem.dump(0, 16) == [0] * 16

    def test_dump_after_write(self) -> None:
        """Dump reflects written bytes."""
        mem = LocalMemory()
        mem.write_byte(0, 0xFF)
        mem.write_byte(1, 0x42)
        d = mem.dump(0, 4)
        assert d[0] == 0xFF
        assert d[1] == 0x42
        assert d[2] == 0
        assert d[3] == 0

    def test_repr(self) -> None:
        """Repr shows size and non-zero count."""
        mem = LocalMemory(size=64)
        assert "64 bytes" in repr(mem)
        assert "0 non-zero" in repr(mem)
        mem.write_byte(0, 1)
        assert "1 non-zero" in repr(mem)

File: code/packages/ruby/gpu_core/test/test_memory.rb
# frozen_string_literal: true

require_relative "test_helper"

module CodingAdventures
  module GpuCore
    # =========================================================================
    # Tests for LocalMemory.
    # =========================================================================
    #
    # Local memory is the GPU core's private scratchpad -- byte-addressable
    # storage for floats that don't fit in registers.

    class TestMemoryConstruction < Minitest::Test
      # Default memory is 4096 bytes.
      def test_default_size
        mem = LocalMemory.new
        assert_equal 4096, mem.size
      end

      # Can create memory with custom size.
      def test_custom_size
        mem = LocalMemory.new(size: 256)
        assert_equal 256, mem.size
      end

      # Cannot create memory with size < 1.
      def test_invalid_size
        err = assert_raises(ArgumentError) { LocalMemory.new(size: 0) }
        assert_match(/positive/, err.message)
      end

      # Memory starts as all zeros.
      def test_initialized_to_zero
        mem = LocalMemory.new(size: 16)
        16.times do |i|
          assert_equal 0, mem.read_byte(i)
        end
      end
    end

    class TestByteAccess < Minitest::Test
      # Write a byte and read it back.
      def test_read_write_byte
        mem = LocalMemory.new
        mem.write_byte(0, 0x42)
        assert_equal 0x42, mem.read_byte(0)
      end

      # Values are masked to 8 bits.
      def test_byte_masking
        mem = LocalMemory.new
        mem.write_byte(0, 0x1FF) # 9 bits
        assert_equal 0xFF, mem.read_byte(0) # truncated to 8
      end

      # Write and read multiple bytes.
      def test_read_write_bytes
        mem = LocalMemory.new
        data = [0x01, 0x02, 0x03, 0x04]
        mem.write_bytes(0, data)
        assert_equal data, mem.read_bytes(0, 4)
      end

      # Reading past memory bounds raises IndexError.
      def test_out_of_bounds_read
        mem = LocalMemory.new(size: 8)
        err = assert_raises(IndexError) { mem.read_byte(8) }
        assert_match(/out of bounds/, err.message)
      end

      # Writing past memory bounds raises IndexError.
      def test_out_of_bounds_write
        mem = LocalMemory.new(size: 8)
        err = assert_raises(IndexError) { mem.write_byte(8, 0) }
        assert_match(/out of bounds/, err.message)
      end

      # Negative addresses are out of bounds.
      def test_negative_address
        mem = LocalMemory.new
        assert_raises(IndexError) { mem.read_byte(-1) }
      end

      # Multi-byte access that extends past the end fails.
      def test_multi_byte_out_of_bounds
        mem = LocalMemory.new(size: 8)
        assert_raises(IndexError) do
          mem.read_bytes(6, 4) # bytes 6,7,8,9 -- 8 and 9 out of bounds
        end
      end
    end

    class TestFloatAccess < Minitest::Test
      # Store and load an FP32 value.
      def test_store_load_fp32
        mem = LocalMemory.new
        value = FpArithmetic.float_to_bits(3.14, FpArithmetic::FP32)
        mem.store_float(0, value)
        result = mem.load_float(0, FpArithmetic::FP32)
        assert_in_delta 3.14, FpArithmetic.bits_to_float(result), 1e-5
      end

      # Store and load an FP16 value (2 bytes).
      def test_store_load_fp16
        mem = LocalMemory.new
        value = FpArithmetic.float_to_bits(1.0, FpArithmetic::FP16)
        mem.store_float(0, value)
        result = mem.load_float(0, FpArithmetic::FP16)
        assert_equal 1.0, FpArithmetic.bits_to_float(result)
      end

      # Store and load a BF16 value (2 bytes).
      def test_store_load_bf16
        mem = LocalMemory.new
        value = FpArithmetic.float_to_bits(2.0, FpArithmetic::BF16)
        mem.store_float(0, value)
        result = mem.load_float(0, FpArithmetic::BF16)
        assert_equal 2.0, FpArithmetic.bits_to_float(result)
      end

      # FP32 store writes exactly 4 bytes.
      def test_fp32_uses_4_bytes
        mem = LocalMemory.new
        value = FpArithmetic.float_to_bits(1.0, FpArithmetic::FP32)
        mem.store_float(0, value)
        raw = mem.read_bytes(0, 4)
        assert_equal 4, raw.length
        refute_equal [0, 0, 0, 0], raw
      end

      # FP16 store writes exactly 2 bytes.
      def test_fp16_uses_2_bytes
        mem = LocalMemory.new
        value = FpArithmetic.float_to_bits(1.0, FpArithmetic::FP16)
        mem.store_float(0, value)
        raw = mem.read_bytes(0, 2)
        assert_equal 2, raw.length
        refute_equal [0, 0], raw
      end

      # Store multiple floats at non-overlapping addresses.
      def test_multiple_floats_at_different_addresses
        mem = LocalMemory.new
        mem.store_ruby_float(0, 1.0)
        mem.store_ruby_float(4, 2.0)
        mem.store_ruby_float(8, 3.0)
        assert_equal 1.0, mem.load_float_as_ruby(0)
        assert_equal 2.0, mem.load_float_as_ruby(4)
        assert_equal 3.0, mem.load_float_as_ruby(8)
      end

      # Store and load a negative float.
      def test_store_negative
        mem = LocalMemory.new
        mem.store_ruby_float(0, -42.5)
        assert_equal(-42.5, mem.load_float_as_ruby(0))
      end

      # Store and load zero.
      def test_store_zero
        mem = LocalMemory.new
        mem.store_ruby_float(0, 0.0)
        assert_equal 0.0, mem.load_float_as_ruby(0)
      end

      # Test store_ruby_float and load_float_as_ruby.
      def test_convenience_methods
        mem = LocalMemory.new
        mem.store_ruby_float(0, 2.71828, FpArithmetic::FP32)
        result = mem.load_float_as_ruby(0, FpArithmetic::FP32)
        assert_in_delta 2.71828, result, 1e-5
      end

      # Loading a float past memory end raises IndexError.
      def test_float_out_of_bounds
        mem = LocalMemory.new(size: 8)
        assert_raises(IndexError) do
          mem.load_float(6, FpArithmetic::FP32) # needs 4 bytes at 6, goes to 10
        end
      end
    end

    class TestMemoryDump < Minitest::Test
      # Dump of fresh memory is all zeros.
      def test_dump_zeros
        mem = LocalMemory.new(size: 16)
        assert_equal [0] * 16, mem.dump(0, 16)
      end

      # Dump reflects written bytes.
      def test_dump_after_write
        mem = LocalMemory.new
        mem.write_byte(0, 0xFF)
        mem.write_byte(1, 0x42)
        d = mem.dump(0, 4)
        assert_equal 0xFF, d[0]
        assert_equal 0x42, d[1]
        assert_equal 0, d[2]
        assert_equal 0, d[3]
      end

      # to_s shows size and non-zero count.
      def test_to_s
        mem = LocalMemory.new(size: 64)
        assert_includes mem.to_s, "64 bytes"
        assert_includes mem.to_s, "0 non-zero"
        mem.write_byte(0, 1)
        assert_includes mem.to_s, "1 non-zero"
      end
    end
  end
end
